### PR TITLE
Feature/spec align

### PR DESCRIPTION
**Pull request by marto83**
This pull request started with a tidy up while I was documenting what is implemented. 
I removed labels where tests didn't implement the exact spec item and also added tests for:

-  G4 
-  RSA7b3
-  RSC1a
-  RSC1b
-  RSA4a2

Bonus test was added for 382

This also includes an fsharp file to list all the implemented test items.

### CODE REVIEW

File: src/IO.Ably.Shared/AuthOptions.cs [hunk 1]
```diff
@@ -89,8 +90,21 @@ public AuthOptions()
         public AuthOptions(string key)
             : this()
         {
-            var apiKey = ApiKey.Parse(key);
-            Key = apiKey.ToString();
+            if (key.IsEmpty())
+            {
+                throw new AblyException(new ErrorInfo("key string cannot be null", 40000));
```
Comment (QuintinWillison) [inline src/IO.Ably.Shared/AuthOptions.cs]: Not sure if I'm interpreting the spec right but should this not also be `40106` per [RSC1b](https://docs.ably.io/client-lib-development-guide/features/#RSC1a)?

**Comment by marto83** [comment]
> Also, the commit message for [5a4d429](https://github.com/ably/ably-dotnet/commit/5a4d429bfa5dd7dc67d2f28fc09f692dc93d7ff4) states 'Implement' but I only see changes to tests. Am I missing something?

My bad there. I didn't really implement anything but just added tests for the spec item.

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSpecs/GeneralConnectionSpecs.cs [hunk 2]
```diff
@@ -18,7 +18,18 @@ public async Task ShouldUseWebSocketTransport()
         {
             var client = await GetConnectedClient(); // The transport is created by the connecting state
 
-            client.ConnectionManager.Transport.GetType().Should().BeAssignableTo<ITransport>();
+            client.ConnectionManager.Transport.Should().NotBeNull();
+        }
+
+        [Fact]
+        [Trait("spec", "G4")]
```
[the remark this answers is not in the document]
Comment (marto83) [reply src/IO.Ably.Tests.Shared/Realtime/ConnectionSpecs/GeneralConnectionSpecs.cs]: RTN2f is specifically for the websocket transport. This particular tests doesn't check it. There are others that do.

File: src/IO.Ably.Shared/AuthOptions.cs (hunk 1 above)
Comment (marto83) [reply src/IO.Ably.Shared/AuthOptions.cs]: You read it right. I've corrected the error code and improved the message.

File: tools/list-test-categories.fsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
+open FSharp
```
[the remark this answers is not in the document]
Comment (marto83) [reply tools/list-test-categories.fsx]: I've added a section to README called spec where I show how to run this script and where to find the results.

[the remark this answers is not in the document]
Comment (QuintinWillison) [reply tools/list-test-categories.fsx]: Nice. Thanks.

**Pull request merged**